### ISSUE TITLE
http-proxy does not allow to connect ho https-proxy

### ISSUE THREAD

**Issue by mal19992** [OpenVPN/openvpn#633]
Currently openvpn --http-proxy option only allows a HTTP proxy, i.e. it directly issues the CONNECT command to http proxy. There is no option to connect to a HTTP proxy not directly, but via TLS/SSL, a so called https-proxy option, a regular HTTP proxy behind TLS/SSL, it is rather easy to setup such with apache or nginx.  This suggested --http-proxy-over-TLS option has two very important benefits:

1. A local ISP cannot be able to eavesdrop the IP of actual VPN server during CONNECT phase and block connection proactively. 
2. Most important. OpenVPN is currently frequently blocked/delayed by some ISPs. This option would allow to pass ISP blocking. A OnenVPN connection with http-proxy option currently gets blocked by ISP with exactly the same logic as used when is connected directly. If OpenVPN connection would be encapsulated in HTTPS connection to the proxy -- it would be much more difficult to identify and block such a connection.

This is a request for improvement. Tested on openvpn-2.6.12 and earlier, none support a HTTP proxy over TLS/SSL. This improvement feature is easy to implement and it does not affect other OpenVPN functionality.

**Comment by cron2** [comment]
> This improvement feature is easy to implement 

Such claims should come with actual code that shows that "it is easy to implement".

**Comment by mal19992** [comment]
I will try to look,but I am not very familiar with OpenVPN codebase

**Comment by ordex** [comment]
I agree with @cron2, however I believe this is an interesting feature and would be nice to have.

**Comment by mal19992** [comment]
I reviewed the concept 

1. The concept works perfectly. I proved it with stunnel by running a separate TLS tunnel with stunnel.conf 
```
#engine = pkcs11
foreground = yes
[service]
#engine = auto
#engineId = pkcs11
client = yes
accept = 127.0.0.5:7999
connect = proxy.xxxxxxxxxx.com:443
debug = 7
# uncomment to enable SSL remote server certificate verification 
#CAfile = /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem
#verify = 2
#verifyChain = yes
#checkHost = proxy.xxxxxxxxxx.com
```
and run it as `stunnel stunnel.conf`
This example creates SSL connection to remote http proxy behind SSL proxy.xxxxxxxxxx.com:443, the unencrypted stream is then available locally at 127.0.0.5:7999. OpenVPN http proxy is now set to 127.0.0.5:7999

The result -- the ISP sees only regular https connection (from stunnel) to proxy.xxxxxxxxxx.com:443, the OpenVPN TCP goes under this TLS link. The ISP does not recognize this traffic as being from OpenVPN and does not throttle it. No problem whatsoever.  This solution is the best to VPN traffic identification https://community.openvpn.net/openvpn/wiki/TrafficObfuscation among several I tried. It really works and is completely separate from OpenVPN internals. All it requires -- run apache http proxy behind SSL. Just one extra line with letsencrypt config.

2. It is very inconvenient to run an extra tunnel. It is much easier to include to  OpenVPN an ability to work with http proxy behind SSL. 
I think the best option would be not a separate proxy setting that I originally described in the first post, but, instead, a single boolean config flag 
`  http_proxy_over_ssl = true`
which makes to open TLS connection before opening http proxy. This is somewhere in 
```
--- a/src/openvpn/socket.c
+++ b/src/openvpn/socket.c
@@ -2068,8 +2068,15 @@ phase2_tcp_client(struct link_socket *sock, struct signal_info *sig_info)
 
         if (sock->http_proxy)
         {
+           int proxy_sd=sock->sd
+           if(sock->http_proxy && sock->http_proxy->options &&  sock->http_proxy->options->http_proxy_over_ssl){
+               // FIXME wrap with SSL
+               // we have connection open to proxy. now wrap it with TLS
+               // proxy_sd= // FIXME open TLS connection to socket sock->sd. The  obtained fd gives an unencrypted stream. Pass it as it were a http proxy socket. 
+           }
+
             proxy_retry = establish_http_proxy_passthru(sock->http_proxy,
-                                                        sock->sd,
+                                                        proxy_sd,
                                                         sock->proxy_dest_host,
                                                         sock->proxy_dest_port,
                                                         sock->server_poll_timeout,
```
It just replaces fd passed to `establish_http_proxy_passthru` in socket.c. With a regular http proxy -- just pass it as is. When http proxy is behind SSL -- it opens SSL connection to proxy destination, then use obtained fd of unencrypted contents is passes  through.

3. Basically a single boolean option `http_proxy_over_ssl` is sufficient for this feature. I would add a few more options affecting the TLS connection to http proxy --- just for convenience (and as a replacement to http basic auth)
```
# Enable http proxy over SSL
http_proxy_over_ssl = true
# SSL host name to verify, do not verify host name if empty.  
http_proxy_over_ssl_server_SNI = proxy.xxxxxxxxxx.com
# Client certificate for proxy auth. Optionally use it instead of basic proxy auth 
http_proxy_over_ssl_client_auth_cert = /etc/client.pem
# http_proxy server SSL certificate to verify locally 
http_proxy_over_ssl_server_cert = /etc/proxyserver.pem
```
Actually only a single option  `http_proxy_over_ssl = true` is important, the other controls are to auth the proxy connection, this is a glorified  "basic auth" for proxy sever.

**Comment by ordex** [comment]
> `http_proxy_over_ssl_CONNECT_Host`

this is not specific to httpS. However, in #635 it was found out that `Host:` should contain the target hostname, no?

**Comment by mal19992** [comment]
A created some draft patch 

[tls_proxy.patch](https://github.com/user-attachments/files/26463013/tls_proxy.patch)

I did not put much attention to non-blocking/IO, I just plain added sleep(3) in few places to get proof of concept. 
It mostly works, when I uncomment this
```
               proxy_sd = SSL_get_fd(ssl);

               printf("SSL handshake completed.\n");

               // ============= BEGIN This is a test
               if(true){ // UNCOMMENT HERE !!!!
                   // Just a test, to demonstarte that SSL works
                   // Send an HTTP request or data
                   char buffer[1024];
                   int nb=sprintf(buffer,"CONNECT %s:%s HTTP/1.0\r\nHost: %s\r\n\r\n",
                                  sock->proxy_dest_host, sock->proxy_dest_port, sock->proxy_dest_host);
                   printf("Sending: %s\n", buffer);
                   SSL_write(ssl, buffer, nb);             

                   sleep(3); // FIXME for non-blocking to work
                   
                   // Receive response
                   int bytes = SSL_read(ssl, buffer, sizeof(buffer) - 1);
                   if (bytes > 0) {
                       buffer[bytes] = '\0';  // Null terminate the buffer
                       printf("Received: %s\n", buffer);
                   }
               }// ============= END This is a test

```
It sends proxy request OK, and get a response, 407 code, auth required 
```
SSL handshake completed.
Sending: CONNECT proxy.xxxxxxxxxx.com:1195 HTTP/1.0
Host: proxy.xxxxxxxxxx.com

Received: HTTP/1.1 407 Proxy Authentication Required
Date: Fri, 03 Apr 2026 15:52:38 GMT
Server: Apache/2.4.66
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
Proxy-Authenticate: Digest realm="AUTH REQUIRED", nonce="wylqTpBOBgA=6dcd40f24877ea626be8f7ccdfe8003d972bf0f6", algorithm=MD5, qop="auth"
Content-Length: 455
Connection: close
Content-Type: text/html; charset=iso-8859-1

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
....
```
Everything is OK, code 407 (auth required, etc).

The problem is when openvpn calls the establish_http_proxy_passthru().
The method `recv_line`  returns an empty line, thus no code 407 is identified. 

@cron2 , @ordex  can you help me with:
1. to modify recv_line so it works OK with wrapped socket. This is a difficult problem for me but probably simple for somebody familiar with the codebase. 
2. technical help with non-blocking I/O cleanup, and resources release, this will help to avoid some trivial bugs
3. verify the 6 options names/type I suggested to add in the https://github.com/user-attachments/files/26463013/tls_proxy.patch 

looking at the "BEGIN This is a test", SSL-wrapped connection works OK, the problem is with   `recv_line` that cannot read a line from it.

**Comment by cron2** [comment]
@mal19992 sorry, no time right now to dig into this.  I assume that the file descriptor you get is mostly useless except for use in `select()`  ("is there new data?"), as you can't just `read()` the data without it going through OpenSSL decryption first.  But no idea how this is done in code, and how it would need to get integrated into OpenVPN.

I'm a packet handling guy ("routing, ifconfig/route setup, ...") not the one who understands these SSL libraries.

**Comment by mal19992** [comment]
@cron2 , I mostly finished it , a few issues left. I completely finished proxy part. The only issue I did not manage to solve, is when the HTTP proxy connection (now over TLS) is established in  proxy.c:establish_socks_proxy_passthru
then the created TCP socket should be wrapped with SSL_read/SSL_write for regular openvpn read/write. I did not find the correct place where to do this final change.
This seems to be in socket.c:link_socket_read_tcp  len = recv(sock->sd, BPTR(&frag), BLENZ(&frag), MSG_NOSIGNAL); 
and socket.c:link_socket_write_tcp link_socket_write_tcp_posix(sock, buf);
but I am not sure. I tried to change these two places, but it does not work. HTTP proxy over  TLS works OK, but proper openvpn read/write does not work yet. I am not sure about proper place where this is implemented.

**Comment by schwabe** [comment]
Since you are trying to wrap the whole connection inside another TLS session, you need to the whole SSL_read/SSL_write magic between streambuf and reading/writing from the socket.
